.github/workflows/main.yml

```
@@ -2,6 +2,10 @@ name: tests
 
 on: [push, pull_request, workflow_dispatch]
 
+concurrency: 
+  group: tests-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
```

My recommendation would be to do [what `pip` does in it's CI suite](https://github.com/pypa/pip/blob/ec8edbf5df977bb88e1c777dd44e26664d81e216/.github/workflows/ci.yml#L19-L21):

```suggestion
concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}
  cancel-in-progress: true
```

Since `run_id` is a unique number for each job run, this wouldn't have much of an effect e.g. with multiple pushes to a single branch.

And I think grouping on `head_ref` would have the effect of cancelling in-progress runs on commits pushed to the default branch, which IMO is undesirable.

---

Makes sense, thanks!